File: src/policy/semantic-guards.ts
import { deobfuscateRecursive, detectShellInBase64Blobs } from '../utils/payload-normalizer.js';
import { preprocessForInjectionMatch } from '../utils/injection-preprocess.js';
import { walkStringLeaves } from './arg-leaf-walker.js';
import { evaluatePathGuard, extractPathArgumentValues } from './path-guard.js';
import { evaluateUrlGuard, extractHttpUrlsFromLeaves } from './url-guard.js';
import type { CallContext, PolicyDecision } from './policy-types.js';

const REPO_ARG_FIELDS = new Set(['repo', 'repository', 'owner']);

const SQL_SENSITIVE_TABLES =
  'accounts|customers|users|credentials|secrets|payments|transactions|admin_users|passwords';

const SQL_EXFIL_PATTERNS: RegExp[] = [
  new RegExp(`\\bselect\\b.+\\bfrom\\b.+\\b(?:${SQL_SENSITIVE_TABLES})\\b`, 'i'),
  new RegExp(`\\bselect\\s+\\*\\s+from\\b.+\\b(?:${SQL_SENSITIVE_TABLES})\\b`, 'i'),
  /\b(?:drop|truncate)\s+(?:table|database)\b/i,
  /\bdelete\s+from\b/i,
  /\bunion\b.+\bselect\b/i,
  /\/\*+\s*union\b/i,
  /\bor\s+['"]?1['"]?\s*=\s*['"]?1['"]?\b/i,
  /\bor\s+1\s*=\s*1\b/i,
  /\b'\s*or\s*'1'\s*=\s*'1/i,
  /\badmin'\s*--/i,
  /\bcase\s+when\b.+\b(?:password|secret|token|credential)/i,
  /\bload_file\s*\(/i,
  /\bsleep\s*\(/i,
  /\bbenchmark\s*\(/i,
  /\binto\s+outfile\b/i,
  /\bextractvalue\s*\(/i,
  /\bupdatexml\s*\(/i,
  /"\$where"\s*:/i,
  /"\$gt"\s*:/i,
  /"\$regex"\s*:/i,
  /"\$ne"\s*:/i,
  /\$where\b/i,
  /\$gt\b/i,
  /\$regex\b/i,
  /\$ne\b/i,
  /__schema\b/i,
  /\bintrospection\b/i,
  /\*\)\s*\(\s*uid\s*=/i,
  /\*\)\s*\(/,
  /admin\)\s*\(&/i,
  /\|\s*\(\s*\|/i,
  /\)\s*\(\s*\|/i,
  /\)\s*\)\s*\(/,
];

const BASE64_SHELL_PATTERNS: RegExp[] = [
  /\bbase64\s+(?:-d|--decode)\b.+\|\s*(?:sh|bash|zsh)\b/i,
  /\|\s*base64\s+(?:-d|--decode)\b.+\|\s*(?:sh|bash|zsh)\b/i,
  /\becho\s+['"]?[A-Za-z0-9+/]{12,}={0,2}['"]?\s*\|\s*base64\s+(?:-d|--decode)\b/i,
  /\bbase64\s+(?:-d|--decode)\b\s*<<<?\s*['"]?[A-Za-z0-9+/]{8,}/i,
];

const POWERSHELL_PATTERNS: RegExp[] = [
  /\bpowershell(?:\.exe)?\b/i,
  /\bpwsh\b/i,
  /-enc(?:odedcommand)?\b/i,
  /\biex\b/i,
];

const SSTI_PATTERNS: RegExp[] = [
  /\{\{/,
  /\$\{/,
  /<%/,
  /#\{/,
];

const MULTILINE_INJECTION_PATTERNS: RegExp[] = [
  /(?:ignore|disregard).{0,120}?(?:instructions|rules|guidelines|directives)/ims,
  /(?:system|assistant)[\s\S]{0,40}:\s*you\s+are/ims,
  /<\|(?:endoftext|im_start|im_end)\|>/i,
];

/** Header / identity spoofing in tool arguments (rate-limit bucket evasion). */
const RATE_IDENTITY_EVASION_PATTERNS: RegExp[] = [
  /x-forwarded-for\s*[:=]\s*[\d.,\s]+/i,
  /x-real-ip\s*[:=]\s*[\d.]+/i,
  /(?:client[_-]?id|sub|tenant[_-]?id)\s*[:=]\s*["']?(?:admin|root|system|spoof)/i,
  /(?:reveal|show|print|disclose).{0,40}(?:rate[- ]?limit|ratelimit).{0,30}key/i,
  /\b(?:unlimited|no[- ]?limit|ignore|disable|bypass).{0,40}(?:rate[- ]?limit|throttle|quota)/i,
  /\b(?:flood|spam|hammer|burst).{0,30}(?:tool[- ]?calls?|requests?|api)/i,
];

/** Token budget / resource exhaustion phrasing in arguments. */
const TOKEN_BUDGET_ABUSE_PATTERNS: RegExp[] = [
  /\b(?:ignore|bypass|exceed|disable|raise).{0,40}(?:token[- ]?budget|max[- ]?tokens?|context[- ]?limit)/i,
  /\b(?:fill|exhaust|maximize).{0,30}(?:context|token|window)/i,
];

/** Log / audit trail injection (CRLF, fake log lines, ANSI escapes). */
const LOG_INJECTION_PATTERNS: RegExp[] = [
  /\r\n[^\n]{0,120}\r\n/,
  /\n(?:INFO|WARN|ERROR|DEBUG|AUDIT|TRACE)\s*:\s*(?:user|admin|override|success)/i,
  /\b(?:INFO|WARN|ERROR|DEBUG|AUDIT|TRACE)\s*:\s*(?:user|admin|override=success)/i,
  /\x1b\[[0-9;]*[A-Za-z]/,
  /(?:forge|spoof|inject|poison).{0,40}(?:audit|log|syslog|trail)/i,
  /\]\s*\(\s*#\s*(?:forge|fake|spoof)/i,
  /%0[aAdD]/i,
];

/** Heuristic: string leaf may be a filesystem path. */
const PATH_LIKE = /(?:^|[\s"'`])(?:~\/|\/|\.\/|\.\.|\\|\.kube|\.ssh|\.env|id_rsa)/i;

function extractFieldValues(args: Record<string, unknown>, fields: Set<string>): string[] {
  const out: string[] = [];
  for (const { path, value } of walkStringLeaves(args)) {
    const key = path.split(/[.[\]]/).filter(Boolean).pop()?.toLowerCase() ?? '';
    if (fields.has(key)) out.push(value);
  }
  return out;
}

function extractPathLikeLeaves(args: Record<string, unknown>): string[] {
  return walkStringLeaves(args)
    .map((l) => l.value)
    .filter((v) => PATH_LIKE.test(v));
}

function githubAllowedRepos(): string[] | null {
  const repos = process.env.GUARDIAN_GITHUB_ALLOWED_REPOS?.split(',').map((s) => s.trim()).filter(Boolean);
  if (repos?.length) return repos;
  const orgs = process.env.GUARDIAN_GITHUB_ALLOWED_ORGS?.split(',').map((s) => s.trim()).filter(Boolean);
  if (orgs?.length) return orgs.map((o) => `${o}/`);
  return null;
}

function repoAllowed(repo: string, allowed: string[]): boolean {
  const r = repo.toLowerCase();
  return allowed.some((a) => {
    const p = a.toLowerCase();
    if (p.endsWith('/')) return r.startsWith(p) || r === p.slice(0, -1);
    return r === p;
  });
}

/**
 * Semantic abuse checks (paths, SQL exfil, GitHub repo scope, PowerShell, SSTI).
 * All guards scan every string leaf via `walkStringLeaves`. Prompt injection on
 * requests is handled in PolicyEngine via `scanToolCallArguments` (full rule set).
 */
export function evaluateSemanticGuards(ctx: CallContext): PolicyDecision | null {
  const args = ctx.arguments ?? {};

  const pathCandidates = [
    ...extractPathArgumentValues(args),
    ...extractPathLikeLeaves(args),
  ];
  const pathCheck = evaluatePathGuard([...new Set(pathCandidates)]);
  if (pathCheck.block) {
    return { action: 'block', rule: 'semantic-path-guard', reason: pathCheck.reason! };
  }

  const urlCandidates = extractHttpUrlsFromLeaves(args);
  const urlCheck = evaluateUrlGuard([...new Set(urlCandidates)], ctx.toolName);
  if (urlCheck.block) {
    return { action: 'block', rule: 'semantic-url-guard', reason: urlCheck.reason! };
  }

  for (const { value } of walkStringLeaves(args)) {
    const decodedSql = deobfuscateRecursive(value);
    for (const pattern of SQL_EXFIL_PATTERNS) {
      if (pattern.test(decodedSql)) {
        return {
          action: 'block',
          rule: 'semantic-sql-guard',
          reason: `SQL/NoSQL/LDAP pattern blocked in tool '${ctx.toolName}'`,
        };
      }
    }
  }

  const argsBlob = deobfuscateRecursive(
    walkStringLeaves(args).map((l) => l.value).join('\n'),
  );
  for (const pattern of BASE64_SHELL_PATTERNS) {
    if (pattern.test(argsBlob)) {
      return {
        action: 'block',
        rule: 'semantic-shell-guard',
        reason: 'Base64-decode piped to shell detected in arguments',
      };
    }
  }
  if (detectShellInBase64Blobs(argsBlob)) {
    return {
      action: 'block',
      rule: 'semantic-shell-guard',
      reason: 'Base64 blob decodes to shell/downloader command in arguments',
    };
  }

  const allowedRepos = githubAllowedRepos();
  if (allowedRepos) {
    for (const repo of extractFieldValues(args, REPO_ARG_FIELDS)) {
      if (!repoAllowed(repo, allowedRepos)) {
        return {
          action: 'block',
          rule: 'semantic-github-guard',
          reason: `GitHub repo '${repo}' not in allowlist`,
        };
      }
    }
  } else {
    for (const repo of extractFieldValues(args, REPO_ARG_FIELDS)) {
      if (/(?:attacker|honeypot|evil|malware|exfil)/i.test(repo)) {
        return {
          action: 'block',
          rule: 'semantic-github-guard',
          reason: `Suspicious GitHub repo target: ${repo}`,
        };
      }
    }
  }

  for (const pattern of POWERSHELL_PATTERNS) {
    if (pattern.test(argsBlob)) {
      return {
        action: 'block',
        rule: 'semantic-powershell-guard',
        reason: 'PowerShell execution pattern detected in arguments',
      };
    }
  }

  for (const pattern of SSTI_PATTERNS) {
    if (pattern.test(argsBlob)) {
      return {
        action: 'block',
        rule: 'semantic-ssti-guard',
        reason: 'Server-side template injection pattern detected in arguments',
      };
    }
  }

  const rawLogBlob = walkStringLeaves(args).map((l) => l.value).join('\n');
  for (const pattern of LOG_INJECTION_PATTERNS) {
    if (pattern.test(rawLogBlob)) {
      return {
        action: 'block',
        rule: 'semantic-log-injection',
        reason: 'Log or audit trail injection pattern in arguments',
      };
    }
  }

  const injectionBlob = preprocessForInjectionMatch(
    walkStringLeaves(args).map((l) => deobfuscateRecursive(l.value)).join('\n'),
  );
  if (injectionBlob.trim()) {
    for (const pattern of MULTILINE_INJECTION_PATTERNS) {
      if (pattern.test(injectionBlob)) {
        return {
          action: 'block',
          rule: 'semantic-prompt-injection',
          reason: 'Multi-line prompt injection pattern in arguments',
        };
      }
    }
    for (const pattern of RATE_IDENTITY_EVASION_PATTERNS) {
      if (pattern.test(injectionBlob)) {
        return {
          action: 'block',
          rule: 'semantic-rate-limit-evasion',
          reason: 'Rate-limit or identity key evasion pattern in arguments',
        };
      }
    }
    for (const pattern of TOKEN_BUDGET_ABUSE_PATTERNS) {
      if (pattern.test(injectionBlob)) {
        return {
          action: 'block',
          rule: 'semantic-token-budget-abuse',
          reason: 'Token budget bypass or exhaustion pattern in arguments',
        };
      }
    }
  }

  return null;
}

File: adversarial-harness/python/policy_engine/semantic_guards.py
"""Semantic guards — mirrors semantic-guards.ts."""

from __future__ import annotations

import os
import re
from typing import Any, Optional

from .arg_walker import walk_string_leaves
from .injection_preprocess import preprocess_for_injection_match
from .normalizer import deobfuscate_recursive, detect_shell_in_base64_blobs
from .path_guard import PATH_ARG_FIELDS, PATH_LIKE, evaluate_path_guard
from .types import CallContext, PolicyDecision
from .url_guard import evaluate_url_guard, extract_http_urls_from_leaves

SQL_SENSITIVE = (
    "accounts|customers|users|credentials|secrets|payments|transactions|admin_users|passwords"
)
SQL_EXFIL_PATTERNS = [
    re.compile(rf"\bselect\b.+\bfrom\b.+\b(?:{SQL_SENSITIVE})\b", re.I),
    re.compile(rf"\bselect\s+\*\s+from\b.+\b(?:{SQL_SENSITIVE})\b", re.I),
    re.compile(r"\b(?:drop|truncate)\s+(?:table|database)\b", re.I),
    re.compile(r"\bdelete\s+from\b", re.I),
    re.compile(r"\bunion\b.+\bselect\b", re.I),
    re.compile(r"/\*+\s*union\b", re.I),
    re.compile(r"\bor\s+['\"]?1['\"]?\s*=\s*['\"]?1['\"]?\b", re.I),
    re.compile(r"\bor\s+1\s*=\s*1\b", re.I),
    re.compile(r"\b'\s*or\s*'1'\s*=\s*'1", re.I),
    re.compile(r"\badmin'\s*--", re.I),
    re.compile(r"\b'\s*or\s*'?\d+'?\s*=\s*'?\d+'?\s*--", re.I),
    re.compile(r"\bor\s+'[^']+'\s*=\s*'[^']+'", re.I),
    re.compile(r"\bcase\s+when\b.+\b(?:password|secret|token|credential)", re.I),
    re.compile(r"\bload_file\s*\(", re.I),
    re.compile(r"\bsleep\s*\(", re.I),
    re.compile(r"\bbenchmark\s*\(", re.I),
    re.compile(r"\binto\s+outfile\b", re.I),
    re.compile(r'"\$where"\s*:', re.I),
    re.compile(r'"\$gt"\s*:', re.I),
    re.compile(r'"\$regex"\s*:', re.I),
    re.compile(r'"\$ne"\s*:', re.I),
    re.compile(r"\$where\b", re.I),
    re.compile(r"\$gt\b", re.I),
    re.compile(r"\$regex\b", re.I),
    re.compile(r"\$ne\b", re.I),
    re.compile(r"__schema\b", re.I),
    re.compile(r"\bintrospection\b", re.I),
    re.compile(r"admin\)\s*\(&", re.I),
    re.compile(r"\*\)\s*\(\s*uid\s*=", re.I),
    re.compile(r"\*\)\s*\(", re.I),
    re.compile(r"\|\s*\(\s*\|", re.I),
    re.compile(r"\)\s*\(\s*\|", re.I),
    re.compile(r"\)\s*\)\s*\(", re.I),
]

BASE64_SHELL_PATTERNS = [
    re.compile(r"\bbase64\s+(?:-d|--decode)\b.+\|\s*(?:sh|bash|zsh)\b", re.I),
    re.compile(r"\|\s*base64\s+(?:-d|--decode)\b.+\|\s*(?:sh|bash|zsh)\b", re.I),
    re.compile(
        r"\becho\s+['\"]?[A-Za-z0-9+/]{12,}={0,2}['\"]?\s*\|\s*base64\s+(?:-d|--decode)\b",
        re.I,
    ),
]

POWERSHELL_PATTERNS = [
    re.compile(r"\bpowershell(?:\.exe)?\b", re.I),
    re.compile(r"\bpwsh\b", re.I),
    re.compile(r"-enc(?:odedcommand)?\b", re.I),
    re.compile(r"\biex\b", re.I),
]

SSTI_PATTERNS = [re.compile(r"\{\{"), re.compile(r"\$\{"), re.compile(r"<%"), re.compile(r"#\{")]

MULTILINE_INJECTION = [
    re.compile(
        r"(?:ignore|disregard).{0,120}?(?:instructions|rules|guidelines|directives)",
        re.I | re.M | re.S,
    ),
    re.compile(r"(?:system|assistant)[\s\S]{0,40}:\s*you\s+are", re.I | re.M | re.S),
    re.compile(r"<\|(?:endoftext|im_start|im_end)\|>", re.I),
]

RATE_IDENTITY_EVASION = [
    re.compile(r"x-forwarded-for\s*[:=]\s*[\d.,\s]+", re.I),
    re.compile(r"x-real-ip\s*[:=]\s*[\d.]+", re.I),
    re.compile(r"(?:client[_-]?id|sub|tenant[_-]?id)\s*[:=]\s*['\"]?(?:admin|root|system|spoof)", re.I),
    re.compile(r"(?:reveal|show|print|disclose).{0,40}(?:rate[- ]?limit|ratelimit).{0,30}key", re.I),
    re.compile(r"\b(?:unlimited|no[- ]?limit|ignore|disable|bypass).{0,40}(?:rate[- ]?limit|throttle|quota)", re.I),
    re.compile(r"\b(?:flood|spam|hammer|burst).{0,30}(?:tool[- ]?calls?|requests?|api)", re.I),
    re.compile(r"\bflood\s+probe\b", re.I),
]

PRIVILEGE_ESCALATION = [
    re.compile(r"\brun\s+as\s+root\b", re.I),
    re.compile(r"\bfull\s+privileges\b", re.I),
    re.compile(r"\belevate\s+privileges\b", re.I),
    re.compile(r"\badministrator\s+with\s+full\b", re.I),
]

DOCKER_ESCAPE = [
    re.compile(r"\bdocker\s+run\b", re.I),
    re.compile(r"\b-v\s+/:/host\b", re.I),
    re.compile(r"\b--privileged\b", re.I),
]

MALICIOUS_HOST_RE = re.compile(
    r"(?:^|[.\-/])(?:evil|attacker|exfil|honeypot|malware)(?:[.\-]|$)|(?:^|[.\-/])evil\.(?:io|com|net|org)\b",
    re.I,
)

PROMPT_OVERRIDE = [
    re.compile(r"\boverride\b.{0,60}\b(?:system|instructions?|prompt|rules?)\b", re.I),
    re.compile(r"\badmin\s+bypass\b", re.I),
    re.compile(r"\b(?:new|unrestricted)\s+assistant\s+persona\b", re.I),
    re.compile(r"\byou\s+are\s+now\s+an?\s+unrestricted\s+assistant\b", re.I),
]

TOKEN_BUDGET_ABUSE = [
    re.compile(r"\b(?:ignore|bypass|exceed|disable|raise).{0,40}(?:token[- ]?budget|max[- ]?tokens?|context[- ]?limit)", re.I),
    re.compile(r"\b(?:fill|exhaust|maximize).{0,30}(?:context|token|window)", re.I),
]

LANGUAGE_GADGET = [
    re.compile(r"\b(?:pickle\.loads?|cPickle|__reduce__|PYCC)\b", re.I),
    re.compile(r"\bObjectInputStream\b|\breadObject\s*\(\s*\)", re.I),
    re.compile(r"\bjava\.io\.(?:Serializable|ObjectInputStream)\b", re.I),
    re.compile(r"\b(?:node-serialize|unserialize\s*\(|__wakeup|__destruct)\b", re.I),
    re.compile(r"\bMarshal\.load\b|\byaml\.unsafe_load\b", re.I),
]

LOG_INJECTION = [
    re.compile(r"\r\n[^\n]{0,120}\r\n"),
    re.compile(r"\n(?:INFO|WARN|ERROR|DEBUG|AUDIT|TRACE)\s*:\s*(?:user|admin|override|success)", re.I),
    re.compile(
        r"\n\[(?:INFO|WARN|ERROR|DEBUG|AUDIT|TRACE)\][\s\S]{0,160}(?:admin|password|secret)",
        re.I,
    ),
    re.compile(
        r"\[(?:INFO|WARN|ERROR|DEBUG|AUDIT|TRACE)\][^\]]{0,120}(?:admin|password|secret)",
        re.I,
    ),
    re.compile(r"\b(?:INFO|WARN|ERROR|DEBUG|AUDIT|TRACE)\s*:\s*(?:user|admin|override=success)", re.I),
    re.compile(r"\x1b\[[0-9;]*[A-Za-z]"),
    re.compile(r"(?:forge|spoof|inject|poison).{0,40}(?:audit|log|syslog|trail)", re.I),
    re.compile(r"\]\s*\(\s*#\s*(?:forge|fake|spoof)", re.I),
    re.compile(r"%0[aAdD]", re.I),
]

REPO_FIELDS = frozenset({"repo", "repository", "owner"})


def _extract_path_candidates(args: dict[str, Any]) -> list[str]:
    out: list[str] = []
    if args:
        for key, val in args.items():
            if key.lower() in PATH_ARG_FIELDS and isinstance(val, str):
                out.append(val)
    out.extend(leaf.value for leaf in walk_string_leaves(args) if PATH_LIKE.search(leaf.value))
    return list(dict.fromkeys(out))


def evaluate_semantic_guards(
    ctx: CallContext,
    raw_arguments: dict[str, Any] | None = None,
) -> Optional[PolicyDecision]:
    args = ctx.arguments or {}
    raw_args = raw_arguments if raw_arguments is not None else args

    path_candidates = _extract_path_candidates(args) + _extract_path_candidates(raw_args)
    path_check = evaluate_path_guard(list(dict.fromkeys(path_candidates)))
    if path_check.block:
        return PolicyDecision("block", "semantic-path-guard", path_check.reason)

    urls = list(
        dict.fromkeys(
            extract_http_urls_from_leaves(args) + extract_http_urls_from_leaves(raw_args),
        ),
    )
    url_check = evaluate_url_guard(urls, ctx.tool_name)
    if url_check.block:
        return PolicyDecision("block", "semantic-url-guard", url_check.reason)

    sql_values: set[str] = set()
    for leaf in walk_string_leaves(args):
        sql_values.add(leaf.value)
    for leaf in walk_string_leaves(raw_args):
        sql_values.add(leaf.value)

    for value in sql_values:
        for candidate in (value, deobfuscate_recursive(value)):
            for pat in SQL_EXFIL_PATTERNS:
                if pat.search(candidate):
                    return PolicyDecision(
                        "block",
                        "semantic-sql-guard",
                        f"SQL/NoSQL/LDAP pattern blocked in tool '{ctx.tool_name}'",
                    )
        if MALICIOUS_HOST_RE.search(value):
            return PolicyDecision(
                "block",
                "semantic-dns-exfil",
                f"Suspicious exfiltration hostname in tool '{ctx.tool_name}'",
            )

    blob = deobfuscate_recursive("\n".join(leaf.value for leaf in walk_string_leaves(args)))
    for pat in BASE64_SHELL_PATTERNS:
        if pat.search(blob):
            return PolicyDecision(
                "block",
                "semantic-shell-guard",
                "Base64-decode piped to shell detected in arguments",
            )
    if detect_shell_in_base64_blobs(blob):
        return PolicyDecision(
            "block",
            "semantic-shell-guard",
            "Base64 blob decodes to shell/downloader command in arguments",
        )

    for pat in POWERSHELL_PATTERNS:
        if pat.search(blob):
            return PolicyDecision(
                "block",
                "semantic-powershell-guard",
                "PowerShell execution pattern detected in arguments",
            )

    for pat in SSTI_PATTERNS:
        if pat.search(blob):
            return PolicyDecision(
                "block",
                "semantic-ssti-guard",
                "Server-side template injection pattern detected in arguments",
            )

    for pat in LANGUAGE_GADGET:
        if pat.search(blob):
            return PolicyDecision(
                "block",
                "semantic-language-gadget",
                "Language-specific deserialization gadget pattern in arguments",
            )

    for pat in PRIVILEGE_ESCALATION:
        if pat.search(blob):
            return PolicyDecision(
                "block",
                "semantic-privilege-escalation",
                "Privilege escalation phrasing detected in arguments",
            )

    for pat in DOCKER_ESCAPE:
        if pat.search(blob):
            return PolicyDecision(
                "block",
                "semantic-shell-guard",
                "Container escape or privileged docker pattern in arguments",
            )

    raw_log = "\n".join(leaf.value for leaf in walk_string_leaves(args))
    for pat in LOG_INJECTION:
        if pat.search(raw_log):
            return PolicyDecision(
                "block",
                "semantic-log-injection",
                "Log or audit trail injection pattern in arguments",
            )

    inj_blob = preprocess_for_injection_match(
        "\n".join(deobfuscate_recursive(leaf.value) for leaf in walk_string_leaves(args))
    )
    if inj_blob.strip():
        for pat in MULTILINE_INJECTION:
            if pat.search(inj_blob):
                return PolicyDecision(
                    "block",
                    "semantic-prompt-injection",
                    "Multi-line prompt injection pattern in arguments",
                )
        for pat in PROMPT_OVERRIDE:
            if pat.search(inj_blob):
                return PolicyDecision(
                    "block",
                    "semantic-prompt-injection",
                    "Prompt or system override pattern in arguments",
                )
        for pat in RATE_IDENTITY_EVASION:
            if pat.search(inj_blob):
                return PolicyDecision(
                    "block",
                    "semantic-rate-limit-evasion",
                    "Rate-limit or identity key evasion pattern in arguments",
                )
        for pat in TOKEN_BUDGET_ABUSE:
            if pat.search(inj_blob):
                return PolicyDecision(
                    "block",
                    "semantic-token-budget-abuse",
                    "Token budget bypass or exhaustion pattern in arguments",
                )
    for leaf in walk_string_leaves(args):
        key = leaf.path.split(".")[-1].lower()
        if key in REPO_FIELDS and re.search(r"(?:attacker|honeypot|evil|malware|exfil)", leaf.value, re.I):
            return PolicyDecision(
                "block",
                "semantic-github-guard",
                f"Suspicious GitHub repo target: {leaf.value}",
            )

    allowed = os.environ.get("MASTYFF_AI_GITHUB_ALLOWED_REPOS", "").strip()
    if allowed:
        allowed_list = [s.strip() for s in allowed.split(",") if s.strip()]
        for leaf in walk_string_leaves(args):
            key = leaf.path.split(".")[-1].lower()
            if key in REPO_FIELDS and not any(
                leaf.value == a or leaf.value.startswith(f"{a}/") for a in allowed_list
            ):
                return PolicyDecision(
                    "block",
                    "semantic-github-guard",
                    f"GitHub repo '{leaf.value}' not in allowlist",
                )

    return None
